tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc

```diff
@@ -139,57 +184,95 @@ bool DotGpuFusionStrategy::initFusionPattern(ShapeAnalysis& shapeAnalysis,
-      if (user == op) {
-        continue;
+  if (transpose != nullptr) {
+    // Check all paths from dom to roots for transpose op.
+    SmallVector<SmallVector<Operation*>> all_paths_dom_to_roots;
+    DenseMap<Operation*, DenseSet<SmallVector<Operation*>>> path_dom_to_ops;
+    Operation* dom = fusion_pattern.getDominantOp();
+    DenseSet<SmallVector<Operation*>> path_to_dom{SmallVector<Operation*>{dom}};
+    path_dom_to_ops.try_emplace(dom, std::move(path_to_dom));
+    SmallVector<Operation*> worklist;
+    worklist.push_back(dom);
+    auto roots = fusion_pattern.getRootOps();
+    DenseSet<Operation*> op_set(op_list.begin(), op_list.end());
+    while (!worklist.empty()) {
+      auto curr = worklist.back();
+      worklist.pop_back();
+      const auto& path_curr = path_dom_to_ops[curr];
+      DenseSet<Operation*> consumers;
+      getDirectConsumerOps(curr, consumers);
+      for (auto consumer : consumers) {
+        if (!op_set.contains(consumer)) {
+          continue;
+        }
+        auto& path_consumer = path_dom_to_ops[consumer];
+        for (const auto& path : path_curr) {
+          SmallVector<Operation*> new_path = path;
+          new_path.push_back(consumer);
+          path_consumer.insert(std::move(new_path));
+        }
+        worklist.push_back(consumer);
+      }
+    }
+    for (auto root : roots) {
+      auto& path = path_dom_to_ops[root];
+      all_paths_dom_to_roots.insert(all_paths_dom_to_roots.end(), path.begin(),
+                                    path.end());
+    }
+    for (auto& path : all_paths_dom_to_roots) {
+      if (llvm::find(path, transpose) == path.end()) {
+        return false;
       }
-      consumers.insert(user);
     }
-  }
-}
-
-void getDirectProducerOpsInFusionPattern(Operation* op,
-                                         FusionPattern& fusion_pattern,
-                                         DenseSet<Operation*>& producers) {
-  producers.clear();
 
-  auto& op_list = fusion_pattern.getOpList();
-  DenseSet<Operation*> op_set(op_list.begin(), op_list.end());
+    // Make sure that the input dims of the transpose op are the same with the
+    // output of the dot op.
+    if (!shapeAnalysis.isShapeEqual(dot_ops[0]->getOperand(2),
+                                    transpose->getOperand(0))) {
+      return false;
+    }
 
-  int64_t num_input_operand = op->getNumOperands() - getNumResultOperands(op);
-  for (Value value : op->getOperands().take_front(num_input_operand)) {
-    auto producer = fusion_pattern.findLastWriter(value);
-    if (op_set.contains(producer)) {
-      producers.insert(producer);
+    // Only support permutation of [0,2,1,3], and dim-1 is static-shaped.
+    auto perm_attr = dyn_cast<lmhlo::TransposeOp>(transpose)
+                         .getPermutation()
+                         .getValues<int64_t>();
+    SmallVector<int64_t> perm{perm_attr.begin(), perm_attr.end()};
+    if (!(perm[0] == 0 && perm[1] == 2 && perm[2] == 1 && perm[3] == 3)) {
```

Comment:
nit: cutlass doesn't support other kind of transpose ?

Reply:
It supports some more but does not appear in standard BERT-like models. It can be extended in the future.